### PR TITLE
Return an error upon fail to obtain pod info

### PR DESCRIPTION
**Pull request by ruigulala**
Fixes: https://github.com/istio/istio/issues/17583
Add a simple wait loop to wait until the pod's info is available from
k8s API. Return the k8s error if it fails eventually.

### CODE REVIEW

File: cmd/istio-cni/main.go [hunk 1]
```diff
@@ -169,9 +171,22 @@ func cmdAdd(args *skel.CmdArgs) error {
...
+			var annotations map[string]string
+			var ports []string
+			var k8sErr error
+			for retry := 0; retry < podRetrievalMaxRetries; retry++ {
+				containers, initContainersMap, _, annotations, ports, k8sErr = getKubePodInfo(client, string(k8sArgs.K8S_POD_NAME), string(k8sArgs.K8S_POD_NAMESPACE))
+				if k8sErr == nil {
+					break
+				}
+				log.Warnf("Waiting for pod data %v", k8sErr)
+				time.Sleep(10 * time.Second)
```
Comment (rlenglet) [inline cmd/istio-cni/main.go]: Define a constant for this.

File: cmd/istio-cni/main.go [hunk 2]
```diff
@@ -169,9 +171,22 @@ func cmdAdd(args *skel.CmdArgs) error {
...
+			var initContainersMap map[string]struct{}
+			var annotations map[string]string
+			var ports []string
+			var k8sErr error
+			for retry := 0; retry < podRetrievalMaxRetries; retry++ {
+				containers, initContainersMap, _, annotations, ports, k8sErr = getKubePodInfo(client, string(k8sArgs.K8S_POD_NAME), string(k8sArgs.K8S_POD_NAMESPACE))
+				if k8sErr == nil {
+					break
+				}
+				log.Warnf("Waiting for pod data %v", k8sErr)
```
Comment (rlenglet) [inline cmd/istio-cni/main.go]:
```suggestion
				log.Warn("Waiting for pod metadata", zap.Error(k8sErr), zap.Int("attempt", attempt))
```

File: cmd/istio-cni/main.go (hunk 1 above)
Comment (rlenglet) [inline cmd/istio-cni/main.go]: This is too long. IIRC, CNI has a timeout, so we should try to return ASAP.
Set this to something like 1 or 2 seconds?

File: cmd/istio-cni/main.go [hunk 3]
```diff
@@ -169,9 +171,22 @@ func cmdAdd(args *skel.CmdArgs) error {
...
+				time.Sleep(10 * time.Second)
+			}
 			if k8sErr != nil {
-				log.Warnf("Error getting Pod data %v", k8sErr)
+				log.Errorf("Failed to get pod data: %v", k8sErr)
```
Comment (rlenglet) [inline cmd/istio-cni/main.go]:
```suggestion
				log.Error("Failed to get pod data", zap.Error(k8sErr))
```

File: cmd/istio-cni/main.go [hunk 4]
```diff
@@ -169,9 +171,22 @@ func cmdAdd(args *skel.CmdArgs) error {
 				return err
 			}
 			log.Debug("Created Kubernetes client", zap.Reflect("client", client))
-			containers, initContainersMap, _, annotations, ports, k8sErr := getKubePodInfo(client, string(k8sArgs.K8S_POD_NAME), string(k8sArgs.K8S_POD_NAMESPACE))
+			var containers []string
+			var initContainersMap map[string]struct{}
+			var annotations map[string]string
+			var ports []string
+			var k8sErr error
+			for retry := 0; retry < podRetrievalMaxRetries; retry++ {
```
Comment (rlenglet) [inline cmd/istio-cni/main.go]:
```suggestion
			for attempt := 1; attempt <= podRetrievalMaxRetries; attempt++ {
```

File: cmd/istio-cni/main.go (hunk 1 above)
Comment (ruigulala) [reply cmd/istio-cni/main.go]: Done.

File: cmd/istio-cni/main.go (hunk 2 above)
Comment (ruigulala) [reply cmd/istio-cni/main.go]: Done.

File: cmd/istio-cni/main.go (hunk 1 above)
Comment (ruigulala) [reply cmd/istio-cni/main.go]: Done. Changed to 1 second.

File: cmd/istio-cni/main.go (hunk 3 above)
Comment (ruigulala) [reply cmd/istio-cni/main.go]: Done.

File: cmd/istio-cni/main.go (hunk 4 above)
Comment (ruigulala) [reply cmd/istio-cni/main.go]: Done.

**Pull request merged**